grails-app/services/org/pih/warehouse/core/IdentifierService.groovy

```diff
@@ -150,13 +153,83 @@ class IdentifierService {
+        def sequenceFormat = ""
+        while (matcher.find()) {
+            sequenceFormat = matcher.group()
+            count++
+        }
+        if (count > allowedSequences) {
+            throw new IllegalArgumentException("Cannot have more sequence numbers than ${allowedSequences} in the same identifier")
+        }
+
+        def sequenceNumber = productType.getNextSequenceNumber()
+        List identifierFormatComponents
+        List identifierComponents
+
+        // Split custom identifier by sequence format (with keeping sequence inside identifier components)
+        String sequenceDelimiter = "((?<=${sequenceFormat})|(?=${sequenceFormat}))"
+        identifierFormatComponents = productType.productIdentifierFormat.split(sequenceDelimiter)
+        identifierComponents = identifierFormatComponents.collect { String identifierFormatComponent ->
+            if (identifierFormatComponent.contains("0")) {
+                return generateSequenceNumber(sequenceNumber.toString(), identifierFormatComponent)
+            }
+            else {
+                return generateIdentifier(identifierFormatComponent)
+            }
+        }
+
+        return identifierComponents.join("")
+    }
+
+    String generateProductIdentifier(ProductType productType) {
+        // If the product type is null, then generate product code from DEFAULT_PRODUCT_NUMBER_FORMAT
+        if (!productType) {
+            return generateProductIdentifier()
+        }
+
+        // If the product type does not have a custom identifier but has code, then generate sequential product code
+        if (!productType.productIdentifierFormat && productType.code) {
+            Integer sequenceNumber = productType.getNextSequenceNumber()
+            String sequenceNumberStr = generateSequenceNumber(sequenceNumber.toString())
+
+            String template = ConfigurationHolder.config.openboxes.identifier.productCode.format
+            String delimiter = ConfigurationHolder.config.openboxes.identifier.productCode.delimiter
+            Map properties = ConfigurationHolder.config.openboxes.identifier.productCode.properties
+            Map model = dataService.transformObject(productType, properties)
+            model.put("sequenceNumber", sequenceNumberStr)
+            model.put("delimiter", delimiter)
+            return renderTemplate(template, model)
+        }
+
+        // If the product type does not have a custom identifier, then generate product code from DEFAULT_PRODUCT_NUMBER_FORMAT
+        if (!productType.productIdentifierFormat) {
+            return generateProductIdentifier()
+        }
+
+        // if does not contain sequential part, then generate identifier basing on custom format
+        if (!productType.productIdentifierFormat.contains("0")) {
+            return generateIdentifier(productType.productIdentifierFormat)
+        }
+
+        return generateSequentialProductIdentifier(productType, Constants.DEFAULT_SEQUENCE_NUMBER_PATTERN, 1)
```

Comment:
Replace the Constants.DEFAULT_SEQUENCE_NUMBER_PATTERN with a configuration property. Each sequential number should have their own pattern added to Config.groovy and the default value should be set to Constants.DEFAULT_SEQUENCE_NUMBER_PATTERN. Actually if the code is getting to this point isn't productType.productIdentifierFormat probably not null? If so I think we should use the pattern in productType.productIdentifierFormat, right?

Reply:
Sorry DEFAULT_SEQUENCE_NUMBER_PATTERN is different from DEFAULT_SEQUENCE_NUMBER_FORMAT.

Reply:
So I guess the case I'm thinking about should be handled by the if above this line so I think we're good.

Reply:
I'll rename the const to DEFAULT_SEQUENCE_NUMBER_FORMAT_CHAR. The productType is passed, because the next sequence number is pulled from it